File: packages/js/src/core/codec/tone.ts
import { applyFade } from "./envelope.js";

export interface ToneSampleOptions {
  sampleRate: number;
  durationMs: number;
  levelDb: number;
  fadeMs: number;
}

export function toneToSamples(freq: number, options: ToneSampleOptions): Float32Array {
  const { sampleRate, durationMs, levelDb, fadeMs } = options;
  const sampleCount = Math.max(1, Math.round((durationMs / 1000) * sampleRate));
  const out = new Float32Array(sampleCount);
  const amplitude = Math.pow(10, levelDb / 20);
  const phaseStep = (2 * Math.PI * freq) / sampleRate;
  let phase = 0;
  for (let i = 0; i < sampleCount; i += 1) {
    phase += phaseStep;
    if (phase > Math.PI * 2) {
      phase -= Math.PI * 2;
    }
    out[i] = Math.sin(phase) * amplitude;
  }
  applyFade(out, sampleRate, fadeMs);
  return out;
}

File: packages/python/qraudio/codec/tone.py
from __future__ import annotations

import math

from .envelope import applyFade


def toneToSamples(
    *,
    freq: float,
    sample_rate: float,
    duration_ms: float,
    level_db: float,
    fade_ms: float,
) -> list[float]:
    sample_count = max(1, round((duration_ms / 1000.0) * sample_rate))
    amplitude = 10 ** (level_db / 20.0)
    phase_step = (2 * math.pi * freq) / sample_rate
    phase = 0.0
    out: list[float] = [0.0] * sample_count
    for i in range(sample_count):
        phase += phase_step
        if phase > math.pi * 2:
            phase -= math.pi * 2
        out[i] = math.sin(phase) * amplitude
    applyFade(out, sample_rate, fade_ms)
    return out
